src/test/ui/lint/special-upper-lower-cases.rs

```diff
@@ -0,0 +1,24 @@
+// (#77273) These characters are in the general categories of
+// "Uppercase/Lowercase Letter".
+// The diagnostics don't provide meaningful suggestions for them
+// as we cannot convert them properly.
+
+// check-pass
+
+#![feature(non_ascii_idents)]
+#![allow(uncommon_codepoints, unused)]
+
+struct 𝕟𝕠𝕥𝕒𝕔𝕒𝕞𝕖𝕝;
+//~^ WARN: type `𝕟𝕠𝕥𝕒𝕔𝕒𝕞𝕖𝕝` should have an upper camel case name
+
+// FIXME: How we should handle this?
+struct 𝕟𝕠𝕥_𝕒_𝕔𝕒𝕞𝕖𝕝;
+//~^ WARN: type `𝕟𝕠𝕥_𝕒_𝕔𝕒𝕞𝕖𝕝` should have an upper camel case name
```

Comment:
@estebank Do you have any ideas to improve this case?

Reply:
we can probably improve this in a separate pr if needed

Reply:
@JohnTitor I just took a look at this. The problem is that `is_lowercase(𝕟)` is true but `to_uppercase(𝕟) == 𝕟`. Changing `char_has_case` to instead of asking whether the char is `lowercase ^ uppercase`, we can do `c.to_lowercase().to_string() != c.to_uppercase().to_string()`. That check would be more reliable to us to decide how to reconstitute the suggestion. But I would go one step further and say that maybe we shouldn't be warning for structs that have a first char that is lowercase but can't be uppercased 🤔

(Yes, I know this is 3 months later. I've had this tab open since then to look into this once I had enough time 😅)

Reply:
what do you think of the suggestion given in https://github.com/rust-lang/rust/pull/77805#discussion_r527821452 : check if the tranformed string would also produce a warning, and don't warn if it does.

Reply:
That could make sense. After all the PascalCase convention for types is good to differentiate them from bindings and functions, but when writing a struct or trait called 𝛅 you probably are encoding "business logic" with prior conventions. It doesn't make that much sense to complain about it.

I am somewhat surprised at the behavior here because although 𝕟 doesn't have a corresponding upper-case character, 𝕞  does: 𝕄, but `to_uppercase` doesn't touch it, it ignores all math symbols and gives you the same symbol, which is why the `to_lowercase(x) != to_uppercase(x)` check would be more resilient.